experimental/dds/tree2/src/class-tree/schemaFactory.ts

```diff
@@ -0,0 +1,578 @@
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils";
+import { RestrictiveReadonlyRecord, fail, getOrCreate, isReadonlyArray } from "../util";
+import {
+	FlexTreeNode,
+	LeafNodeSchema as FlexLeafNodeSchema,
+	isFlexTreeNode,
+	ObjectNodeSchema,
+	isLazy,
+	markEager,
+} from "../feature-libraries";
+import { leaf } from "../domains";
+import { TreeValue } from "../core";
+import { TreeMapNodeBase } from "../simple-tree";
+import {
+	createNodeProxy,
+	createRawObjectProxy,
+	getClassSchema,
+	getSequenceField,
+	listPrototypeProperties,
+	mapStaticDispatchMap,
+	// eslint-disable-next-line import/no-internal-modules
+} from "../simple-tree/proxies";
+import { getFlexSchema, setFlexSchemaFromClassSchema } from "./toFlexSchema";
+import {
+	AllowedTypes,
+	FieldKind,
+	FieldSchema,
+	ImplicitAllowedTypes,
+	ImplicitFieldSchema,
+	InsertableObjectFromSchemaRecord,
+	InsertableTreeNodeFromImplicitAllowedTypes,
+	InsertableTypedNode,
+	NodeBase,
+	NodeFromSchema,
+	NodeKind,
+	ObjectFromSchemaRecord,
+	TreeListNode,
+	TreeNodeFromImplicitAllowedTypes,
+	TreeNodeSchema,
+	TreeNodeSchemaClass,
+	TreeNodeSchemaNonClass,
+} from "./schemaTypes";
+
+/**
+ * Instances of this class are schema for leaf nodes.
+ * @remarks
+ * Unlike other schema, leaf schema are class instances instead of classes themselves.
+ * This is because the instance type (the tree node type) for leaves are not objects,
+ * so those instances can't be instances of a schema based class.
+ * @privateRemarks
+ * This class referees to the underlying flex tree schema in its constructor, so this class can't be included in the package API.
+ */
+class LeafNodeSchema<T extends FlexLeafNodeSchema>
+	implements
+		TreeNodeSchemaNonClass<T["name"], NodeKind.Leaf, T["leafValue"], TreeValue<T["info"]>>
+{
+	public readonly identifier: T["name"];
+	public readonly kind = NodeKind.Leaf;
+	public readonly info: T["info"];
+	public create(data: TreeValue<T["info"]>): TreeValue<T["info"]> {
+		return data;
+	}
+
+	public constructor(schema: T) {
+		setFlexSchemaFromClassSchema(this, schema);
+		this.identifier = schema.name;
+		this.info = schema.info;
+	}
+}
+
+/**
+ * Wrapper around LeafNodeSchema's constructor that provides the return type that is desired in the package public API.
+ */
+function makeLeaf<T extends FlexLeafNodeSchema>(
+	schema: T,
+): TreeNodeSchema<T["name"], NodeKind.Leaf, T["leafValue"], TreeValue<T["info"]>> {
+	return new LeafNodeSchema(schema);
+}
+
+/**
+ * Builds schema libraries, and the schema within them.
+ *
+ * @typeParam TScope - Scope added as a prefix to the name of every schema produced by this factory.
+ * @typeParam TName - Type of names used to identify each schema produced in this factory.
+ * Typically this is just `string` but it is also possible to use `string` or `number` based enums if you prefer to identify your types that way.
```

Comment:
Using string based enums is possible and potentially useful if you want to centrally manage your identifiers somewhere, and still use strings (ex: whiteboard could make their id files string enums)

Reply:
Oh I see, this is just a grammar ambiguity. I read the comment as "use string OR number-based-enums" rather than "use string-or-number based enums". So I thought the string part was a mistake since you already talked about strings.

Reply:
Would be nice to make it less ambigious, but I don't see a good way to clarify it simply.